src/Host/Client/Impl/Session/RSession.cs

```diff
@@ -215,27 +215,33 @@ internal sealed class RSession : IRSession, IRCallbacks {
         }
 
         public async Task EnsureHostStartedAsync(RHostStartupInfo startupInfo, IRSessionCallback callback, int timeout = 3000, CancellationToken cancellationToken = default(CancellationToken)) {
-            using (_disposeToken.Link(ref cancellationToken))
-            using (await _initializationLock.WaitAsync(cancellationToken)) {
-                if (_initializationTcs.Task.Status != TaskStatus.RanToCompletion || !_isHostRunning) {
-                    await StartHostAsyncBackground(startupInfo, callback, timeout, cancellationToken);
+            using (_disposeToken.Link(ref cancellationToken)) {
+                await TaskUtilities.SwitchToBackgroundThread();
+
+                using (await _initializationLock.WaitAsync(cancellationToken)) {
+                    if (_initializationTcs.Task.Status != TaskStatus.RanToCompletion || !_isHostRunning) {
+                        await StartHostAsyncBackground(startupInfo, callback, timeout, cancellationToken);
+                    }
                 }
             }
         }
 
         public async Task StartHostAsync(RHostStartupInfo startupInfo, IRSessionCallback callback, int timeout = 3000, CancellationToken cancellationToken = default(CancellationToken)) {
-            using (_disposeToken.Link(ref cancellationToken))
-            using (await _initializationLock.WaitAsync(cancellationToken)) {
-                if (_initializationTcs.Task.Status != TaskStatus.RanToCompletion || !_isHostRunning) {
-                    await StartHostAsyncBackground(startupInfo, callback, timeout, cancellationToken);
-                } else {
-                    throw new InvalidOperationException("Another instance of RHost is running for this RSession. Stop it before starting new one.");
+            using (_disposeToken.Link(ref cancellationToken)) {
+                await TaskUtilities.SwitchToBackgroundThread();
+
```

Comment:
Yes, that is why I fixed it so `await TaskUtilities.SwitchToBackgroundThread();` is called before lock is acquired.